packages/dom-adapters/src/BlockToolAdapter/index.ts

```diff
@@ -133,40 +195,82 @@ export class BlockToolAdapter {
       case InputType.DeleteSoftLineForward:
       case InputType.DeleteWordBackward:
       case InputType.DeleteWordForward: {
-        this.#model.removeText(this.#blockIndex, key, start, end);
-
+        if (nativeInput) {
+          this.#handleDeleteInNativeInput(event, input as HTMLInputElement | HTMLTextAreaElement, key);
+        } else {
+          this.#handleDeleteInContentEditable(event, input, key);
+        }
         break;
       }
 
       default:
     }
-  }
+  };
 
-  /**
-   * Handles model update events and updates DOM
-   *
-   * @param event - model update event
-   * @param input - attched input element
-   * @param key - data key input is attached to
-   * @param caretAdapter - caret adapter instance
-   */
-  #handleModelUpdate(event: ModelEvents, input: HTMLElement, key: DataKey, caretAdapter: CaretAdapter): void {
+  #handleModelUpdateNative = (event: ModelEvents, input: HTMLInputElement | HTMLTextAreaElement, key: DataKey): void => {
+    if (!(event instanceof TextAddedEvent) && !(event instanceof TextRemovedEvent)) {
+      return;
+    }
+
+    const [, dataIndex, blockIndex] = event.detail.index;
+
+    if (blockIndex !== this.#blockIndex) {
+      return;
+    }
+
+    if (dataIndex !== composeDataIndex(key)) {
+      return;
+    }
+
+    const currentElement = input as HTMLInputElement | HTMLTextAreaElement;
+    const [ [start, end] ] = event.detail.index;
+
+    const action = event.detail.action;
+
+    switch (action) {
+      case EventAction.Added: {
+        const text = event.detail.data as string;
+        const prevValue = currentElement.value;
+
+        currentElement.value = prevValue.slice(0, start) + text + prevValue.slice(end - 1);
+
+        currentElement.setSelectionRange(start + text.length, start + text.length);
+        break;
+      }
+      case EventAction.Removed: {
+        if (
+          start === end &&
```

Comment:
I don't think we can get this case from the model update. If something is deleted, start would not be equal to end